### PR TITLE
CLOUDSTACK-9280: System VM volumes can be expunged if no SSVM exists.

### PR DESCRIPTION
**Pull request by ProjectMoon**
This commit adds a special SSVM endpoint which simply returns true for
all operations sent to it, without actually doing anything. This
allows for destroyed volumes of system VMs to be expunged when there
are no hosts (and thus no system VMs) remaining to handle the volume
destruction.

### CODE REVIEW

**Comment by ProjectMoon** [comment]
New version of #1406, opened against master instead of the 4.6 branch.

**Comment by swill** [comment]
### CI RESULTS

```
Tests Run: 85
  Skipped: 0
   Failed: 1
   Errors: 1
 Duration: 10h 41m 20s
```

**Summary of the problem(s):**
```
FAIL: Test destroy(expunge) Virtual Machine
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data/git/cs1/cloudstack/test/integration/smoke/test_vm_life_cycle.py", line 646, in test_09_expunge_vm
    self.assertEqual(list_vm_response,None,"Check Expunged virtual machine is in listVirtualMachines response")
AssertionError: Check Expunged virtual machine is in listVirtualMachines response
----------------------------------------------------------------------
Additional details in: /tmp/MarvinLogs/test_vpc_routers_NJRMT6/results.txt
```

```
ERROR: Test to verify access to loadbalancer haproxy admin stats page
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data/git/cs1/cloudstack/test/integration/smoke/test_internal_lb.py", line 854, in tearDown
    raise Exception("Cleanup failed with %s" % e)
Exception: Cleanup failed with Job failed: {jobprocstatus : 0, created : u'2016-05-24T12:25:52+0200', jobresult : {errorcode : 530, errortext : u'Failed to delete network'}, cmd : u'org.apache.cloudstack.api.command.user.network.DeleteNetworkCmd', userid : u'fb39f9f4-2168-11e6-9284-5254001daa61', jobstatus : 2, jobid : u'a7e1d97d-e770-43f6-83a7-2d81ada1521e', jobresultcode : 530, jobresulttype : u'object', jobinstancetype : u'Network', accountid : u'fb39dea8-2168-11e6-9284-5254001daa61'}
----------------------------------------------------------------------
Additional details in: /tmp/MarvinLogs/test_network_HC1DJD/results.txt
```



**Associated Uploads**

**`/tmp/MarvinLogs/DeployDataCenter__May_24_2016_06_38_46_2TM014:`**
* [dc_entries.obj](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1559/tmp/MarvinLogs/DeployDataCenter__May_24_2016_06_38_46_2TM014/dc_entries.obj)
* [failed_plus_exceptions.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1559/tmp/MarvinLogs/DeployDataCenter__May_24_2016_06_38_46_2TM014/failed_plus_exceptions.txt)
* [runinfo.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1559/tmp/MarvinLogs/DeployDataCenter__May_24_2016_06_38_46_2TM014/runinfo.txt)

**`/tmp/MarvinLogs/test_network_HC1DJD:`**
* [failed_plus_exceptions.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1559/tmp/MarvinLogs/test_network_HC1DJD/failed_plus_exceptions.txt)
* [results.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1559/tmp/MarvinLogs/test_network_HC1DJD/results.txt)
* [runinfo.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1559/tmp/MarvinLogs/test_network_HC1DJD/runinfo.txt)

**`/tmp/MarvinLogs/test_vpc_routers_NJRMT6:`**
* [failed_plus_exceptions.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1559/tmp/MarvinLogs/test_vpc_routers_NJRMT6/failed_plus_exceptions.txt)
* [results.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1559/tmp/MarvinLogs/test_vpc_routers_NJRMT6/results.txt)
* [runinfo.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1559/tmp/MarvinLogs/test_vpc_routers_NJRMT6/runinfo.txt)


Uploads will be available until `2016-07-25 02:00:00 +0200 CEST`

*Comment created by [`upr comment`](https://github.com/cloudops/upr).*

**Comment by bvbharatk** [comment]
### ACS CI BVT Run
 **Sumarry:**
 Build Number 186
 Hypervisor xenserver
 NetworkType Advanced
 Passed=68
 Failed=5
 Skipped=3

_Link to logs Folder (search by build_no):_ https://www.dropbox.com/sh/yj3wnzbceo9uef2/AAB6u-Iap-xztdm6jHX9SjPja?dl=0


**Failed tests:**
* test_vpc_vpn.py

 * ContextSuite context=TestRVPCSite2SiteVpn>:setup Failing since 19 runs

 * ContextSuite context=TestVpcRemoteAccessVpn>:setup Failing since 19 runs

 * ContextSuite context=TestVpcSite2SiteVpn>:setup Failing since 19 runs

* test_volumes.py

 * test_06_download_detached_volume Failed

* test_vm_life_cycle.py

 * test_10_attachAndDetach_iso Failed


**Skipped tests:**
test_vm_nic_adapter_vmxnet3
test_static_role_account_acls
test_deploy_vgpu_enabled_vm

**Passed test suits:**
test_deploy_vm_with_userdata.py
test_affinity_groups_projects.py
test_portable_publicip.py
test_over_provisioning.py
test_global_settings.py
test_scale_vm.py
test_service_offerings.py
test_routers_iptables_default_policy.py
test_routers.py
test_reset_vm_on_reboot.py
test_snapshots.py
test_deploy_vms_with_varied_deploymentplanners.py
test_login.py
test_list_ids_parameter.py
test_public_ip_range.py
test_multipleips_per_nic.py
test_regions.py
test_affinity_groups.py
test_network_acl.py
test_pvlan.py
test_nic.py
test_deploy_vm_root_resize.py
test_resource_detail.py
test_secondary_storage.py
test_disk_offerings.py

**Comment by ProjectMoon** [comment]
Updated to latest master.

**Comment by ProjectMoon** [comment]
This pull request has been updated with another commit that fixes a bug where volumes cannot be expunged because the loading of the volume fails to find a primary data store (despite the expunger being able to handle this scenario). Will be working on some unit tests for this.

**Comment by ProjectMoon** [comment]
The OOBM power off test keeps failing. I've run the test locally with the simulator and it succeeds fine. Going to try to kick Travis again to see if it works (third time's the charm?).

**Comment by ProjectMoon** [comment]
Looks like Travis worked this time.

File: engine/storage/test/org/apache/cloudstack/storage/endpoint/DefaultEndPointSelectorTest.java [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+    /**
+     * Dummy endpoint should be selected if an SSVM volume is expunging
+     * and no SSVMs are available.
+     */
+    @Test
+    public void testSelectionWithExpungingSsvm() {
+        EndPoint ep = mockEndpointSelection(Type.SecondaryStorageVm, State.Expunging);
+        Assert.assertNotNull(ep);
+        Assert.assertTrue(ep instanceof DummyEndpoint);
+        Assert.assertEquals(ep, dummyEndpoint);
```
Comment (GabrielBrascher) [inline engine/storage/test/org/apache/cloudstack/storage/endpoint/DefaultEndPointSelectorTest.java]: Lines 114 - 116 could be extracted to a method like the following:

```
private EndPoint executeTestSelectionWithExpunging(Type vmType, Object returnedEndPoint) {
    EndPoint ep = mockEndpointSelection(vmType, State.Expunging);
    Assert.assertNotNull(ep);
    Assert.assertEquals(ep, returnedEndPoint);
    return ep;
}
```

The endpoint returned by _executeTestSelectionWithExpunging(Type vmType, Object equalsTo)_ can be used in methods that perform additional tests (e.g. `Assert.assertTrue(ep instanceof DummyEndpoint);`).

**Comment by rhtyd** [comment]
@ProjectMoon can you fix the conflicts, thanks

**Comment by ProjectMoon** [comment]
Updated to latest 4.8. Will also address the comments too.

**Comment by ProjectMoon** [comment]
Addressed the comments. For some reason the tests for VolumeDataFactoryImpl were being skipped. Will try to see why.

**Comment by rhtyd** [comment]
Thanks @ProjectMoon can you squash your commits.
@murali-reddy @abhinandanprateek can you help review this?

**Comment by ProjectMoon** [comment]
@rhtyd, in my opinion the two commits are functionally separate enough that they should remain separate (and this might have been asked for in one of the various previous incarnations of this PR--I don't remember). I think rolling them into one commit puts too many things in one change. The smaller commits allows reverting of functionality easier, etc. If the change is rebased/fast-forward merged, then it shouldn't be any different than if there were two PRs.

**Pull request closed without merging**